### DISCUSSION TITLE
Nested destructuring problem with prod build

### DISCUSSION THREAD

**Discussion by gautam-jha** [vercel/next.js#27301] [Help]
### What version of Next.js are you using?

11

### What version of Node.js are you using?

16.4.2, 14.17.3

### What browser are you using?

Chrome

### What operating system are you using?

Linux

### How are you deploying your application?

Vercel

### Describe the Bug

 Nested object properties become undefined during production build, but works fine with the dev environment.

```JS
const { featured: { data: { posts } } } = props;
```

The workaround that I am using right now for this.

```JS 
const posts = props?.featured?.data?.posts;
```

![image](https://user-images.githubusercontent.com/16630278/126030746-a5b81da0-05ad-4add-bd03-0c6e2846e52f.png)

### Expected Behavior

Getting a property posts from props.

### To Reproduce

```JS
const { featured: { data: { posts } } } = props;
```

![image](https://user-images.githubusercontent.com/16630278/126030888-876292f2-280c-4b79-9e43-ee7c9b5f1738.png)

**Comment by mattcarlotta** [comment]
### Problem

Unrelated to Next, but this is the result of some (asynchronous) undefined properties being deconstructed without initially being defined.

### Example

Demo (read [comments](https://codesandbox.io/s/vigorous-feynman-fhn5l?fontsize=14&hidenavigation=1&theme=dark&file=/src/App.js:269-484)):
[![Edit Deconstruct Error Before Defined](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/deconstruct-error-before-defined-fhn5l?fontsize=14&hidenavigation=1&theme=dark)

Code (see comments):
```jsx
import * as React from "react";
import "./styles.css";

export default function App() {
  const [state, setState] = React.useState({});
  // this works because deconstruction is only going one level deep (each deconstructed property is initially undefined)
  const { error, data, isLoading } = state;

  // this throws an error because it's trying to deconstruct properties from "data," which is initially undefined
  const { completed, id, title } = data;

  React.useEffect(() => {
    const fetchData = async () => {
      try {
        const res = await fetch("https://jsonplaceholder.typicode.com/todos/1");
        if (res.status !== 200) throw String("Unable to locate todo item");
        const data = await res.json();

        setState({ error: "", data, isLoading: false });
      } catch (error) {
        setState({ error: error.toString(), data: {}, isLoading: false });
      }
    };

    fetchData();
    /* eslint-disable react-hooks/exhaustive-deps */
  }, []);

  return (
    <div className="App">
      <h1>Hello CodeSandbox</h1>
      <h2>Start editing to see some magic happen!</h2>
      {error ? (
        <p style={{ color: "red" }}>{error}</p>
      ) : isLoading ? (
        <p>Loading...</p>
      ) : (
        <>
          <div>
            <strong>Id</strong>: {id}
          </div>
          <div>
            <strong>Title</strong>: {title}
          </div>
          <div>
            <strong>Completed</strong>: {completed.toString()}
          </div>
        </>
      )}
    </div>
  );
}
```

In short, you can't deconstruct a property from an **undefined** object.

### Solution

Either define the `data` properties:

```jsx
  const [state, setState] = React.useState({
     error: "",
     data: { completed: false, id: "", title: "" },
     isLoading: true
   });
   const { error, data: { completed, id, title }, isLoading } = state;

  ...fetchData
  
  return (...);
```

Or conditionally render the `data` properties:
```jsx
  const [state, setState] = React.useState({
     error: "",
     data: {},
     isLoading: true
   });
   const { error, data, isLoading } = state;

   ...fetchData

   return (
     Object.keys(data).length > 0 
       ?  <>
            <div>Id: {data.id}</div>
            <div>Title: {data.title}</div>
            <div>Completed: {data.completed.toString()}</div>
          </>
       : null
   );
```

Optionally, you can make sure that each property within `data` is defined -- this is overkill if you already know before-hand that when `data` is defined, then the properties: `id`, `title` and `completed` will always be defined:
```jsx
  return (
    Object.keys(data).length > 0 && data.id && data.title && data.completed
       ?  <>
            <div>Id: {data.id}</div>
            <div>Title: {data.title}</div>
            <div>Completed: {data.completed.toString()}</div>
          </>
       : null
  );

**Comment by gautam-jha** [comment]
@mattcarlotta Thanks, I guess I choose the wrong method of passing data to child.
 data is coming from `getStaticProps` which is further passed to a child component, where deconstruct didn't worked for me.
 
I also think it is not related to Next but some error with implementation
